bin/generate_peptides.py

```diff
@@ -157,23 +158,21 @@ def main(args=None):
                 results = results.groupby(["protein_id", "peptide_sequence"]).size().reset_index(name="count")
                 # -> protein_id, peptide_sequence, count
                 results["count"] = pd.to_numeric(results["count"], downcast="unsigned")
-                # prepare df for joining
-                results.set_index("peptide_sequence", inplace=True)
-                results.sort_index(inplace=True)
 
-                unique_peptides = pd.DataFrame(index=results.index.drop_duplicates())
-                unique_peptides["peptide_id"] = range(id_counter, id_counter + len(unique_peptides))
-                id_counter += len(unique_peptides)
+                results = results.sort_values(by="peptide_sequence")
+                pep_ids = results.groupby("peptide_sequence").ngroup()
+                results["peptide_id"] = pep_ids + id_counter
+                id_counter = id_counter + len(pep_ids)
+
                 # -> peptide_sequence, peptide_id
-                unique_peptides.to_csv(pep_handle, mode="a", sep="\t", index=True, header=print_header)
+                results[["peptide_id", "peptide_sequence"]].drop_duplicates().sort_values(by=["peptide_sequence","peptide_id"]).to_csv(pep_handle, mode="a", sep="\t", index=False, header=print_header)
 
-                results = results.join(unique_peptides)
                 # -> protein_id, peptide_sequence, count, peptide_id
 
                 print("\nInfo: results (['protein_id','peptide_sequence','peptide_id','count'])", flush=True)
                 results.info(verbose=False, memory_usage=print_mem)
 
-                results[["protein_id", "peptide_id", "count"]].to_csv(
+                results[["protein_id", "peptide_id", "count"]].drop_duplicates().sort_values(by=["protein_id", "peptide_id"]).to_csv(
```

Comment:
Is it necessary to change it that much? Just would aim at minimal changes to avoid introducing new issues.
Is the protein order chaning? Then one could sort maybe already here https://github.com/nf-core/metapep/blob/c126e2776e52e0ca63110cd5fc98c8fe3a89571b/bin/generate_peptides.py#L146

Reply:
I don't remember the details, if you think it makes sense to change it, its ok of course ;)

Reply:
I had the feeling it helps readability to use only one data frame instead of generating a second one for peptides + ids which is then joined back to the results. I think you could be right sorting it already could make it deterministic, but I'm not 100% sure as the peptides are extracted by the amino acids additionally to the proteins, so I'm not sure if it actually works! I would keep it like that.